### ISSUE TITLE
Reproducing Table 1

### ISSUE THREAD

**Issue by zhenduow** [amzn/cosearcher#3]
Hi,

Thank you for your work and sharing the code!

I am trying to reproduce the table 1 results with my own code. I used the [repo](https://github.com/aliannejadi/qulac) to run the QL retriever and get the output.

My question is how to get the retrieval performance after appending the clarifying question and generated answer to the query. 

I find it not so straightforward because the input of QL is a query_id. How to get the query_id after appending context? Or is there anything I am missing?

Thank you!

**Comment by alexandres** [comment]
Hi!

Thank you for your interest in our work. 🤗 

Can you have a look at https://github.com/amzn/cosearcher/blob/main/src/ir.py#L73 and see if it answers your question?

Best,
Alex

**Comment by zhenduow** [comment]
Thank you! I believe I get it now. Thanks for making that really easy to use.

**Comment by zhenduow** [comment]
Hi alexandres, 

I find that a coorperative answer generator is mentioned in the paper. However, in the code, these no-answers are not generated, instead they are from the dataset. Can you confirm that is just how it is supposed to work? 

Thanks!

**Comment by alexandres** [comment]
Yes, that is correct. See https://github.com/amzn/cosearcher/blob/main/src/answer_generation.py

If you want, you can create your own answer generator as described here https://github.com/amzn/cosearcher#customization

In our work, we just use the `similarity` argument (which describes the match between question and user facet) in deciding whether to answer yes/no, but if you want to be more flexible in answering questions, the `AnswerGenerator.generate_answer` abstract method should be changed to accept as input the question that was asked.

**Comment by zhenduow** [comment]
> Yes, that is correct. See https://github.com/amzn/cosearcher/blob/main/src/answer_generation.py
> 
> If you want, you can create your own answer generator as described here https://github.com/amzn/cosearcher#customization
> 
> In our work, we just use the `similarity` argument (which describes the match between question and user facet) in deciding whether to answer yes/no, but if you want to be more flexible in answering questions, the `AnswerGenerator.generate_answer` abstract method should be changed to accept as input the question that was asked.

Thank you! It is crystal clear now!

**Comment by zhenduow** [comment]
Hi alexandres,

I have a question: 

I customize a answer generator and a question generator.

The question generator will randomly choose **one question from the questions_answers list** of the facet, and the answer generator will simply return the **facet** as the answer.

I only run this for 1 turn and use the metric_calculator to get the query retrieval performance (the ranking metrics) with these additional information.

However, I find that this has the exact same performance as returning "" (nothing) as the answer. 

I wonder if this is normal. Can you please share some insights?

**Comment by alexandres** [comment]
Hi @zhenduow! I'm happy to help. 

Can you share the code changes you made and the json file that `main.py` outputs?

**Comment by alexandres** [comment]
But I think I can answer your question without looking at your code. If you have a look at our [paper](https://scholar.harvard.edu/files/malmasi/files/salleetal-ecir2021-cosearcher.pdf), you'll see that there is a component called the "Response Classifier". The heuristic used is very simple (see [here](https://github.com/amzn/cosearcher/blob/main/src/yes_no_detection.py)). Since your answer generator is simply returning the facet, the Response Classifier considers the answer a "No" and continues to ask questions. Since it never receives an answer it classifies as  "Yes", it simply uses the topic as the query to the IR system. This is why you're seeing the same performance as if you simply returned a "" answer.

Consider writing your own Response Classifier by subclassing [YesNoDetector](https://github.com/amzn/cosearcher/blob/main/src/yes_no_detection.py#L17) and passing it [here](https://github.com/amzn/cosearcher/blob/main/src/main.py#L108). 

To test your case, always have your `YesNoDetector` return a "yes", and you should see much higher IR numbers when returning facets as answers, as compared to returning "".

**Comment by zhenduow** [comment]
Hi @alexandres,

Thank you very much! I did miss the `YesNoDetector` part. 

So, if I understand correctly, the clarifying question and answer will only be appended to the search query if the `YesNoDetector` returns a "yes", otherwise it always use the query-only mode. Is that right?

I guess I can set the `YesNoDetector` following your suggestion and that should work as I expect.

Thanks again!

**Comment by zhenduow** [comment]
It does give me much higher performance. Thank you!

> But I think I can answer your question without looking at your code. If you have a look at our [paper](https://scholar.harvard.edu/files/malmasi/files/salleetal-ecir2021-cosearcher.pdf), you'll see that there is a component called the "Response Classifier". The heuristic used is very simple (see [here](https://github.com/amzn/cosearcher/blob/main/src/yes_no_detection.py)). Since your answer generator is simply returning the facet, the Response Classifier considers the answer a "No" and continues to ask questions. Since it never receives an answer it classifies as "Yes", it simply uses the topic as the query to the IR system. This is why you're seeing the same performance as if you simply returned a "" answer.
> 
> Consider writing your own Response Classifier by subclassing [YesNoDetector](https://github.com/amzn/cosearcher/blob/main/src/yes_no_detection.py#L17) and passing it [here](https://github.com/amzn/cosearcher/blob/main/src/main.py#L108).
> 
> To test your case, always have your `YesNoDetector` return a "yes", and you should see much higher IR numbers when returning facets as answers, as compared to returning "".

It does improve the performance. Thank you!
